File: python/sorting/common/selection_sort.py
# %%
def selection_sort(arr, reverse=False):
  """Function for selection sort. The algorithm works to sort a list by
  comparing one item with the remaining items in the list and swap that item
  with the smallest one. Continue with the second item until the last one.

  Args:
      arr (list): An unsorted list of non-negative integers
      reverse (bool, optional): False means sorting the list in an increasing order.
      True means sorting the list in a decreasing order.

  Returns:
      arr (list): A sorted array
  """
  step_count = 0
  for i in range(len(arr)):
    min_idx = i
    step_count += 1

    for j in range(i+1, len(arr)-1):
      step_count += 1
      
      if arr[i] > arr[j]:
        step_count += 1
        min_idx = j

    arr[i], arr[min_idx] = arr[min_idx], arr[i]
    step_count += 1
  return arr, step_count

File: javascript/sorting/selection_sort.js
function selectionSort(array) {
  for (let i = 0; i < array.length; i++) {
    let min_idx = i

    for (let j = i + 1; j < array.length; j++) {
      if (array[min_idx] > array[j]) {
        min_idx = j;
      }
    }
    [array[i], array[min_idx]] = [array[min_idx], array[i]]
  }
  return array;
}

console.log(selectionSort([1,4,2,8,345,123,43,32,5643,63,123,43,2,55,1,234,92]))